### PR TITLE
feat(mtmd): add SMT backend multimodal inference support for llama-server and llama-mtmd-cli

### PR DESCRIPTION
**Pull request by co-seven**
## Summary

  This PR adds multimodal large-model inference feature support on the Spacemit SMT backend, targeting `llama-server` and `llama-mtmd-cli` in the `spacemit-mtmd` branch.

  The main goal is to enable SMT-based multimodal inference workflows on Spacemit platforms, including server-side and CLI-side integration, runtime backend selection, model/config compatibility updates, and related usability improvements for
image/audio pipelines.

  ## What is included

  - Add SMT backend multimodal integration for `llama-server`
  - Add SMT backend multimodal integration for `llama-mtmd-cli`
  - Support runtime backend selection and related config handling
  - Add SMT image/audio processing path and warmup support
  - Add profiling support for vision/audio encoder stages
  - Add timing summary for vision/audio encode stages
  - Add SMT-only multimodal chat controls such as `vision_history`
  - Improve backend loading and thread/config handling
  - Add compatibility updates for related multimodal model paths and configs
  - Add support updates for:
    - Qwen3.5-related model arch handling
    - Qwen3-ASR SMT audio path
    - PaddleOCR-VL SMT vision pipeline
  - Add tokenizer/config fallback and encoder input-size compatibility fixes
  - Update build and usage documentation for SMT / multimodal workflows

  ## User-facing impact

  This PR enables multimodal inference feature support on the SMT backend for Spacemit targets, including:

  - image understanding through SMT vision integration
  - audio-capable multimodal workflows on supported models
  - SMT-backed server inference through `llama-server`
  - SMT-backed CLI inference through `llama-mtmd-cli`
  - improved deployment/configuration usability for multimodal inference on Spacemit platforms

  ## Notes

  - This PR focuses on SMT backend multimodal feature enablement and related integration work.
  - The submitted changes are the selected feature/fix commits prepared for upstreaming into `spacemit-mtmd`.

### CODE REVIEW

File: tools/server/server-context.cpp [hunk 1]
```diff
@@ -449,20 +474,42 @@ struct server_slot {
...
-                t_prompt_processing + t_token_generation, n_prompt_tokens_processed + n_decoded);
+        // text-only tokens (excluding media tokens counted in image/audio prefill)
+        const int32_t n_text_tokens = n_prompt_tokens_processed - n_vision_tokens - n_audio_tokens;
+        // text prefill time = total prompt time minus media prefill time
+        const double t_text_prefill_ms = t_prompt_processing - t_vision_prefill_ms - t_audio_prefill_ms;
+
+        const double t_text_per_tok  = n_text_tokens > 0 ? t_text_prefill_ms / n_text_tokens : 0.0;
+        const double n_text_per_sec  = n_text_tokens > 0 ? 1e3 / t_text_prefill_ms * n_text_tokens : 0.0;
+        const double t_gen_per_tok   = t_token_generation / n_decoded;
+        const double n_gen_per_sec   = 1e3 / t_token_generation * n_decoded;
```
[the remark this answers is not in the document]
Comment (co-seven) [reply tools/server/server-context.cpp]: fixed

File: ggml/src/ggml-profile.c [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+    }
+
+    pthread_mutex_lock(&g_trace_mutex);
+
+    if (g_trace_file) {
+        ggml_trace_tls_flush_locked();
+
+        long pos = ftell(g_trace_file);
+        if (pos >= 2) {
+            fseek(g_trace_file, pos - 2, SEEK_SET);  // remove ",\n"
```
[the remark this answers is not in the document]
Comment (co-seven) [reply ggml/src/ggml-profile.c:278]: skip

File: ggml/src/ggml-profile.c [hunk 3]
```diff
@@ -0,0 +1,245 @@
+#include "ggml-profile.h"
+
+#include <pthread.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/syscall.h>
+#include <time.h>
+#include <unistd.h>
```
[the remark this answers is not in the document]
Comment (co-seven) [reply ggml/src/ggml-profile.c]: fixed

File: convert_hf_to_gguf.py [hunk 4]
```diff
@@ -1250,15 +1250,69 @@ def does_token_look_special(self, token: str | bytes) -> bool:
...
+            logger.warning(
+                "AutoTokenizer failed for %s with tokenizer_class=%s; "
+                "falling back to AutoProcessor.tokenizer",
+                self.dir_model,
+                tokenizer_class,
+            )
+
+            from transformers import AutoProcessor
+
+            processor = AutoProcessor.from_pretrained(self.dir_model, trust_remote_code=True)
```
[the remark this answers is not in the document]
Comment (co-seven) [reply convert_hf_to_gguf.py]: skip

File: tools/mtmd/mtmd-cli.cpp [hunk 5]
```diff
@@ -287,6 +305,13 @@ int main(int argc, char ** argv) {
         return 1;
     }
 
+#if defined(LLAMA_SERVER_SMT_VISION)
+    if (params.media_backend == "smt" || !params.smt_config_dir.empty()) {
```
[the remark this answers is not in the document]
Comment (co-seven) [reply tools/mtmd/mtmd-cli.cpp]: fixed

File: tools/mtmd/smt-audio-wrapper.cpp [hunk 6]
```diff
@@ -0,0 +1,813 @@
+#include "smt-audio-wrapper.h"
+
+#include "ggml-profile.h"
+#include "mtmd-audio.h"
+#include "onnxruntime_cxx_api.h"
+#include "onnxruntime_session_options_config_keys.h"
+
+#include <dirent.h>
+#include <dlfcn.h>
```
[the remark this answers is not in the document]
Comment (co-seven) [reply tools/mtmd/smt-audio-wrapper.cpp]: fixed

File: ggml/src/ggml-profile.c [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+
+static inline int64_t ggml_trace_now_us(void) {
+    struct timespec ts;
+    clock_gettime(CLOCK_MONOTONIC, &ts);
+    return (int64_t) ts.tv_sec * 1000000LL + ts.tv_nsec / 1000;
+}
+
+static inline unsigned long ggml_trace_tid(void) {
+    return (unsigned long) syscall(SYS_gettid);
+}
```
[the remark this answers is not in the document]
Comment (co-seven) [reply ggml/src/ggml-profile.c:77]: done

File: tools/mtmd/mtmd-cli-smt.cpp [hunk 8]
```diff
@@ -0,0 +1,1408 @@
...
+
+        if (!ctx.tok_img_end.empty()) {
+            if (decode_tokens(ctx.lctx, ctx.tok_img_end, ctx.n_past, ctx.n_batch, logits_last) != 0) {
+                LOG_ERR("Failed to decode image-end token chunk %zu\n", i);
+                return 1;
+            }
+        }
+    }
+
+    ctx.pending_media.clear();
```
[the remark this answers is not in the document]
Comment (co-seven) [reply tools/mtmd/mtmd-cli-smt.cpp]: done

File: tools/server/server-smt-vision.h [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+
+inline int32_t server_smt_vision_decode_chunk(
+        llama_context * /* lctx */,
+        const server_smt_vision_context * /* ctx */,
+        const server_smt_image_chunk & /* chunk */,
+        llama_pos & /* n_past */,
+        int32_t /* seq_id */,
+        int32_t /* n_batch */,
+        bool /* logits_last */) {
+    return 1;
```
[the remark this answers is not in the document]
Comment (co-seven) [reply tools/server/server-smt-vision.h]: done

File: convert_hf_to_gguf.py [hunk 10]
```diff
@@ -1250,15 +1250,69 @@ def does_token_look_special(self, token: str | bytes) -> bool:
...
         tokens: list[str] = []
         toktypes: list[int] = []
 
-        from transformers import AutoTokenizer
-        tokenizer = AutoTokenizer.from_pretrained(self.dir_model)
-        vocab_size = self.hparams.get("vocab_size", len(tokenizer.vocab))  # ty: ignore[unresolved-attribute]
-        assert max(tokenizer.vocab.values()) < vocab_size  # ty: ignore[unresolved-attribute]
+        tokenizer = self.load_hf_tokenizer()
+        vocab_size = self.hparams.get("vocab_size", len(tokenizer.vocab))
+        assert max(tokenizer.vocab.values()) < vocab_size
```
[the remark this answers is not in the document]
Comment (co-seven) [reply convert_hf_to_gguf.py]: done
